sdk/communication/azure-communication-chat/src/main/java/com/azure/android/communication/chat/implementation/notifications/fcm/RegistrationKeyManager.java

```diff
@@ -0,0 +1,435 @@
+import android.security.keystore.KeyProperties;
+import android.util.Pair;
+
+import androidx.annotation.NonNull;
+
+import com.azure.android.core.logging.ClientLogger;
+import com.azure.android.core.util.Base64Util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * A singleton wrapper which manage the persistence and rotation of secrete keys and execution result related to #{RenewalTokenWorker}.
+ * Secret keys information is persisted in two files. One is the Android-platform provided key-store API for actual secret key. The
+ * other is the alias with the creation time.
+ *
+ * During each refresh operation. A new pair of secret keys are generated as cryptoKey and authKey. They are persisted into the file
+ * system. While secret keys created longer than #{EXPIRATION_TIME_MINUTES} get removed.
+ */
+public final class RegistrationKeyManager {
+    private static RegistrationKeyManager registrationKeyManager;
+
+    private KeyGenerator androidKeyGenerator;
+
+    private KeyGenerator secretKeyGenerator;
+
+    private KeyStore keyStore;
+
+    private KeyMetaDataStore keyMetaDataStore;
+
+    private EnCryptor enCryptor;
+
+    private DeCryptor deCryptor;
+
+    private boolean lastExecutionSucceeded = true;
+
+    //The duration we persist keys in key-store
+    public static final int EXPIRATION_TIME_MINUTES = 45;
+
+    public static final String CRYPTO_KEY_PREFIX = "CRYPTO_KEY_";
+
+    public static final String AUTH_KEY_PREFIX = "AUTH_KEY_";
+
+    public static final String KEY_CREATION_TIME_POSTFIX = "/key-creation-time-store";
+
+    private ClientLogger clientLogger = new ClientLogger(RegistrationKeyManager.class);
+
+    private RegistrationKeyManager()  {
+        enCryptor = new EnCryptor();
+        deCryptor = new DeCryptor();
+        try {
+            this.secretKeyGenerator = KeyGenerator.getInstance("AES");
+            this.secretKeyGenerator.init(256);
+            androidKeyGenerator = KeyGenerator.getInstance(
+                KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
```

Comment:
I noticed we are using a secondary `KeyGenerator` with a different implementation than the one by "AndroidKeyStore". From what I can tell below, it's so that we can access its contents via `getEncoded()`. Is this so that the key's entire contents can be sent to the service?

Reply:
Yes. We could directly send the keys generated by secretKeyGenerator to registration service. The content should be the same as what decrypted from the file. I implement it in a way to send the decrypted content, for easier debugging. It makes it easier to validate the validation of encryption/decryption process. I removed the logging logic in PR to avoid leaking the content on client side.